### PR TITLE
Add more types

### PR DESCRIPTION
**Pull request by montyly**
Results of monkeytype on the AST tests + manual edits

The PR is not ready for review, but I want to test the CI before doing more changes

### CODE REVIEW

**Comment by montyly** [comment]
Still a lot of types are missing, but the PR is ready to be merged

File: slither/core/declarations/function.py [hunk 1]
```diff
@@ -1711,7 +1718,13 @@ def _unchange_phi(ir: "Operation"):
             return True
         return ir.rvalues[0] == ir.lvalue
 
-    def fix_phi(self, last_state_variables_instances, initial_state_variables_instances):
+    def fix_phi(
+        self,
+        last_state_variables_instances: Dict[
```
Comment (0xalpharush) [inline slither/core/declarations/function.py:1723]: Can this be simplified to `List[Union[StateIRVariable", StateIRVariable"]]`

File: slither/core/solidity_types/array_type.py [hunk 2]
```diff
@@ -1,13 +1,24 @@
...
+    from slither.core.solidity_types.elementary_type import ElementaryType
+    from slither.core.solidity_types.function_type import FunctionType
+    from slither.core.solidity_types.type_alias import TypeAliasTopLevel
 
 
 class ArrayType(Type):
-    def __init__(self, t, length):
+    def __init__(
+        self,
+        t: Union["TypeAliasTopLevel", "ArrayType", "FunctionType", "ElementaryType"],
```
Comment (0xalpharush) [inline slither/core/solidity_types/array_type.py:19]: This doesn't seem right

File: slither/core/solidity_types/array_type.py [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+        length: Optional[Union["Identifier", Literal, "BinaryOperation"]],
```
Comment (0xalpharush) [inline slither/core/solidity_types/array_type.py:20]: This can be int

File: slither/core/solidity_types/user_defined_type.py [hunk 4]
```diff
@@ -8,10 +8,14 @@
...
     from slither.core.declarations.enum import Enum
     from slither.core.declarations.contract import Contract
+    from slither.core.declarations import EnumContract
+    from slither.core.declarations.structure_top_level import StructureTopLevel
 
 # pylint: disable=import-outside-toplevel
 class UserDefinedType(Type):
-    def __init__(self, t):
+    def __init__(
+        self, t: Union["EnumContract", "StructureTopLevel", "Contract", "StructureContract"]
```
Comment (0xalpharush) [inline slither/core/solidity_types/user_defined_type.py:17]: There's some confusion here about whether it's a Structure or Union[StructureTopLevel, StructureContract] between the type annotations and assertions

File: slither/detectors/compiler_bugs/array_by_reference.py [hunk 5]
```diff
@@ -83,7 +87,15 @@ def get_funcs_modifying_array_params(contracts):
...
     @staticmethod
-    def detect_calls_passing_ref_to_function(contracts, array_modifying_funcs):
+    def detect_calls_passing_ref_to_function(
+        contracts: List[Contract], array_modifying_funcs: Set[FunctionContract]
+    ) -> List[
+        Union[
+            Tuple[Node, StateVariable, FunctionContract],
+            Tuple[Node, LocalVariable, FunctionContract],
+            Any,
+        ]
```
Comment (0xalpharush) [inline slither/detectors/compiler_bugs/array_by_reference.py:97]: Can we remove the Any?

File: slither/detectors/erc/erc20/incorrect_erc20_interface.py [hunk 6]
```diff
@@ -36,7 +40,9 @@ class IncorrectERC20InterfaceDetection(AbstractDetector):
     )
 
     @staticmethod
-    def incorrect_erc20_interface(signature):
+    def incorrect_erc20_interface(
+        signature: Union[Tuple[str, List[str], List[Any]], Tuple[str, List[Any], List[Any]]]
```
Comment (0xalpharush) [inline slither/detectors/erc/erc20/incorrect_erc20_interface.py:44]: f.signature is defined as `Tuple[str, List[str]]` so perhaps that can be sued

File: slither/detectors/shadowing/local.py [hunk 7]
```diff
@@ -50,7 +57,18 @@ class LocalShadowing(AbstractDetector):
...
     OVERSHADOWED_EVENT = "event"
 
-    def detect_shadowing_definitions(self, contract):  # pylint: disable=too-many-branches
+    # pylint: disable=too-many-branches
+    def detect_shadowing_definitions(
+        self, contract: Contract
+    ) -> List[
+        Union[
+            Any,
+            Tuple[LocalVariable, List[Tuple[str, StateVariable]]],
```
Comment (0xalpharush) [inline slither/detectors/shadowing/local.py:66]: Can this be simplified to a Union on the last second element of the tuple?

File: slither/slithir/convert.py [hunk 8]
```diff
@@ -300,7 +322,7 @@ def _make_function_type(func: Function) -> FunctionType:
 ###################################################################################
 
 
-def integrate_value_gas(result):
+def integrate_value_gas(result: List[Any]) -> List[Any]:
```
Comment (0xalpharush) [inline slither/slithir/convert.py:325]: I believe should be Operation

File: slither/slithir/convert.py [hunk 9]
```diff
@@ -1114,7 +1141,7 @@ def extract_tmp_call(ins: TmpCall, contract: Optional[Contract]):  # pylint: dis
 ###################################################################################
 
 
-def can_be_low_level(ir):
+def can_be_low_level(ir: slither.slithir.operations.high_level_call.HighLevelCall) -> bool:
```
Comment (0xalpharush) [inline slither/slithir/convert.py:1144]: Is there a reason to use this over just `HighLevelCall`?

File: slither/visitors/slithir/expression_to_slithir.py [hunk 10]
```diff
@@ -146,10 +167,10 @@ def __init__(self, expression, node):  # pylint: disable=super-init-not-called
         for ir in self._result:
             ir.set_node(node)
 
-    def result(self):
+    def result(self) -> List[Any]:
```
Comment (0xalpharush) [inline slither/visitors/slithir/expression_to_slithir.py:170]: This should probably be Operation

**Pull request merged**